docs/rfcs/027-pageserver-disaster-recovery.md

```diff
@@ -0,0 +1,198 @@
+  historic versions beyond the range of history that we officially support.
+* Running the recovery process in the background for random tenant timelines
+  to figure out if there was a corruption of data (we would compare with what
+  the pageserver stores for the "official" timeline).
+* Using the WAL to arrive at historic pages we can then back up to S3 so that
+  WAL itself can be discarded, or at least not used for future replays.
+  Again, this sounds a lot like what the pageserver is already doing, but the
+  point is to provide a fallback to the service provided by the pageserver.
+
+## Design
+
+### Design constraints
+
+The main design constraint is that the feature needs to be *simple* enough that
+the number of bugs are as low, and reliability as high as possible: the main
+goal of this endeavour is to achieve higher correctness than the pageserver.
+
+For the background process, we cannot afford a downtime of the timeline that is
+being cloned, as we don't want to restrict ourselves to offline tenants only.
+In the scenario where we want to recover from disasters or roll back to a
+historic lsn through support staff, downtimes are more affordable, and
+inevitable if the original had been subject to the corruption. Ideally, the
+two code paths would share code, so the solution would be designed for not
+requiring downtimes.
+
+### API endpoints
+
+The proposed design of the new pageserver API endpoint is:
+
+```
+/v1/tenant/:tenant_id/timeline/:timeline_id/recover?from_lsn=<from_lsn>&to_lsn=<lsn>&target=<target_timeline_id>
+```
+
+The `recover` endpoint answers to `POST` requests and issues a WAL recovery
+process. It takes the WAL of the specified timeline, and writes its results
+into a *different* timeline, the target timeline. This two timeline scheme
+ensures that we can stil serve requests and preserves immutability guarantees
+assumed by the [generation numbers RFC], as in corruption situations we might
+indeed end up with different content than the original timeline we are storing
+data from.
+
+Where the parameters are:
+
+* `:tenant_id`: the id of the tenant to recover
+* `:timeline_id`: the id of the timeline to recover
+* `from_lsn`: The lsn at the start of our recovery process.
+  Optional, if not specified the recovery process is fully WAL based.
+  If specified, the recovery process starts from data at the *main* timeline
+  at the specific lsn (maybe we still trust that lsn). In any case, the target
+  timeline must be less advanced than `from_lsn`.
+* `to_lsn`: the end of the lsn range we want to recover. As usual, it's an
+  exclusive limit. This allows playing back the WAL to a specific moment in the
+  past, if this is wanted/needed. E.g. there might be a data-corrupting
+  pageserver crash at lsn 1234 and we don't want that so we play back until
+  1233.
+* `target`: the timeline id we are writing into. The parameter is optional;
+  if no target timeline is provided, a new timeline is created.
```

Comment:
it's just the way of specifying the ID of the new timeline. alternatively, one could rename `target` to `source` and make `:timeline_id` be the ID of the newly created timeline. I'm fine with either, please tell me which one you like more.

But yeah, the pageserver will not expect any WAL pertaining to lsns from the specified range to come from safekeepers for the newly created timeline (target). Instead, it takes the WAL from the source timeline. Also from the safekeepers.

Reply:
> it's just the way of specifying the ID of the new timeline. alternatively, one could rename target to source and make :timeline_id be the ID of the newly created timeline. I'm fine with either, please tell me which one you like more.

I'm also fine with either, `target` is ok. But I'd rephrase description that we always create a new timeline, and `target` should not match any existing timeline. If not specified, it will be generated randomly.